src/OrderedMsgChain.js

```diff
@@ -0,0 +1,108 @@
+import debugFactory from 'debug'
+import Heap from 'heap'
+import { Errors, MessageLayer } from 'streamr-client-protocol'
+
+const debug = debugFactory('StreamrClient::OrderedMsgChain')
+const { MessageRef } = MessageLayer
+
+const DEFAULT_GAPFILL_TIMEOUT = 5000
+const MAX_GAP_REQUESTS = 10
+
+export default class OrderedMsgChain {
+    constructor(publisherId, msgChainId, inOrderHandler, gapHandler, gapFillTimeout = DEFAULT_GAPFILL_TIMEOUT) {
+        this.publisherId = publisherId
+        this.msgChainId = msgChainId
+        this.inOrderHandler = inOrderHandler
+        this.gapHandler = gapHandler
+        this.lastReceivedMsgRef = null
+        this.gapFillTimeout = gapFillTimeout
+        this.queue = new Heap((msg1, msg2) => {
+            return msg1.getMessageRef().compareTo(msg2.getMessageRef())
+        })
+    }
+
+    add(unorderedStreamMessage) {
+        const msgRef = unorderedStreamMessage.getMessageRef()
+        if (this.lastReceivedMsgRef && msgRef.compareTo(this.lastReceivedMsgRef) < 0) {
+            // Prevent double-processing of messages for any reason
+            debug('Already received message: %o, lastReceivedMsgRef: %d. Ignoring message.', msgRef, this.lastReceivedMsgRef)
+        }
+
+        if (this._isNextMessage(unorderedStreamMessage)) {
+            this._process(unorderedStreamMessage)
+        } else {
+            if (!this.gap) {
+                this._scheduleGap()
+            }
+            this._insertInOrderedQueue(unorderedStreamMessage)
+        }
+    }
+
+    addError(err) {
+        if (err.streamMessage) {
+            if (err instanceof Errors.InvalidJsonError && this._isNextMessage(err.streamMessage)) {
+                this.lastReceivedMsgRef = err.streamMessage.getMessageRef()
+            }
+        }
+    }
+
+    clearGap() {
+        clearInterval(this.gap)
+        this.gap = undefined
+    }
+
+    _isNextMessage(unorderedStreamMessage) {
+        const isFirstMessage = this.lastReceivedMsgRef === null
+        return isFirstMessage
+            || (this.lastReceivedMsgRef !== null
+                && unorderedStreamMessage.prevMsgRef !== null
```

Comment:
What if the messages are NOT chained (i.e. `unorderedStreamMessage.prevMsgRef` is `null`). These messages should always be considered to be a valid next message (unless they are older than the newest processed message and fail the `compareTo` check, which is done earlier).

Is there test coverage for unchained messages?